http/src/main/scala/org/http4s/blaze/http/http20/Http20FrameDecoder.scala

```diff
@@ -0,0 +1,280 @@
+package org.http4s.blaze.http.http20
+
+import java.nio.ByteBuffer
+
+import Settings.Setting
+import Http2Exception._
+
+import scala.collection.mutable.ArrayBuffer
+
+
+/* The job of the Http20FrameCodec is to slice the ByteBuffers. It does
+   not attempt to decode headers or perform any size limiting operations */
+trait Http20FrameDecoder {
+  import bits._
+
+  def handler: FrameHandler
+
+  /** Decode a data frame. false signals "not enough data" */
+
+  def decodeBuffer(buffer: ByteBuffer): Http2Result = {
+    if (buffer.remaining() < HeaderSize) {
+      return BufferUnderflow
+    }
+
+    buffer.mark()
+    val len: Int = ((buffer.get() & 0xff) << 16) |
+                   ((buffer.get() & 0xff) << 8 ) |
+                    (buffer.get() & 0xff)
+
+    if (len + 6 > buffer.remaining()) {   // We still don't have a full frame
+      buffer.reset()
+      return BufferUnderflow
+    }
+
+    // we have a full frame, get the header information
+    val frameType = buffer.get()
+    val flags = buffer.get()
+    val streamId = buffer.getInt() & Masks.STREAMID // this concludes the 9 byte header. `in` is now to the payload
+    
+    // Make sure we are not in the middle of some header frames
+    if (handler.inHeaderSequence() && frameType != FrameTypes.CONTINUATION) {
+      return Error(PROTOCOL_ERROR(s"Received frame type $frameType while in in headers sequence"))
+    }
+
+    // set frame sizes in the ByteBuffer and decode
+    val oldLimit = buffer.limit()
+    val endOfFrame = buffer.position() + len
+    buffer.limit(endOfFrame)
+
+    val r = frameType match {
+      case FrameTypes.DATA          => decodeDataFrame(buffer, streamId, flags)
+      case FrameTypes.HEADERS       => decodeHeaderFrame(buffer, streamId, flags)
+      case FrameTypes.PRIORITY      => decodePriorityFrame(buffer, streamId, flags)
+      case FrameTypes.RST_STREAM    => decodeRstStreamFrame(buffer, streamId)
+      case FrameTypes.SETTINGS      => decodeSettingsFrame(buffer, streamId, flags)
+      case FrameTypes.PUSH_PROMISE  => decodePushPromiseFrame(buffer, streamId, flags)
+      case FrameTypes.PING          => decodePingFrame(buffer, streamId, flags)
+      case FrameTypes.GOAWAY        => decodeGoAwayFrame(buffer, streamId)
+      case FrameTypes.WINDOW_UPDATE => decodeWindowUpdateFrame(buffer, streamId)
+      case FrameTypes.CONTINUATION  => decodeContinuationFrame(buffer, streamId, flags)
+        
+        // this concludes the types established by HTTP/2.0, but it could be an extension
+      case code                     => onExtensionFrame(code, streamId, flags, buffer.slice())
+    }
+
+    // reset buffer limits
+    buffer.limit(oldLimit)
+    buffer.position(endOfFrame)
+
+    return r
```

Comment:
Is the concern that its a ```return``` statement at the end of the block instead of the expression form? I kept it return just to be consistent with the preceding code: this is really java written in scala. All the checks would have made this a rats nest of ```if/else``` statements. Perhaps it could be broken down into multiple functions, but I thought it nice to keep the code that mutates the buffer as localized as possible.